### ISSUE TITLE
Need Localtunnel Server with .Net (C#)

### ISSUE THREAD

**Issue by zobayerhossainmamun** [angelobreuer/localtunnel.net#12]
I got the localtunnel server with nodejs. But, I need localtunnel server with c#.

**Comment by angelobreuer** [comment]
Hello @zobayerhossainmamun , see #11.

If there is really a need for a .NET implementation, and if there are multiple users that could benefit from a .NET implementation, then it could be worth to implement a .NET version.

Do you have any use cases where a .NET implementation can provide something that the NodeJS implementation is unable to do? Also I would have to ask the owner of the NodeJS server, because he got the main idea. I think then we could also provide WebSocket support. Note that the NodeJS server is used in public and hosted on a server, that the owner has to pay for. I do not have the resources to provide a public-hosted server.

**Comment by angelobreuer** [comment]
cc @savage8811

**Comment by zobayerhossainmamun** [comment]
Actually, I'm using dot net client version. So, If I got server also with dotnet. It's better for me. I don't need to use nodejs with this.

**Comment by angelobreuer** [comment]
I see. I've collected a couple of things that a new .NET implementation could improve over the NodeJS version:
- Fast non-blocking asynchronous server I/O
- High throughput scenarios
- Interface for DNS provisioning that is compatible with existent .NET libraries
- Better threading capabilities than the node-based implementation
- WebSocket Support
- Better interoperability with .NET applications

I will create an issue in the home repo of the localtunnel server, to ask the owner whether he would like to have a .NET based server implementation.

**Comment by zobayerhossainmamun** [comment]
Okay Thanks.

**Comment by zobayerhossainmamun** [comment]
It's very helpful for us if you are able to public .Net version soon.

**Comment by angelobreuer** [comment]
There is an alpha version available at https://github.com/angelobreuer/localtunnel-client/tree/feat/server/src/Localtunnel.Server

The appsettings.json should look like this:

```json
{
  "Dns": {
    "Zone": "localtunnel.example.com"
  },
  "PortAllocation": {
    "BindAddress": "Any" // 'IPv4Only', 'IPv6Only', 'Any' (Dual Stack)
  }
}
```

The latest Visual Studio preview or .NET SDK 6.0.200 is required for compilation.

**Comment by zobayerhossainmamun** [comment]
Thanks a lot for your help.

**Comment by zobayerhossainmamun** [comment]
I'm surprised for your quick update. We are small developers easily can learn more for those developer like you.
I'm really thankful to you. May god gives you more power for nice work.

**Comment by savage8811** [comment]
Wow! Really nice, TY @angelobreuer

**Comment by angelobreuer** [comment]
I am currently working on a full overhaul of the client to support WebSockets, HTTP/1.1, and HTTP/2.0. The server should now be ready for HTTP/1.1, HTTP/2.0, HTTP/3.0, but not WebSockets. I also reworked the passthrough and "non-passthrough" modes. We now support full request parsing using Kestrel's HTTP parser. I think in the future localtunnel could support HTTP+QUIC (aka HTTP/3.0). Now it is more easy to access the tunnel API for server-side ASP.NET (for example setting up a tunnel and changing headers, then passing it to ASP.NET).

I think as this is no longer a "localtunnel-client", this repository will be renamed to "localtunnel.net". The NuGet package name will stay the same.

**Comment by samirsommaior** [comment]
Congratulations on your work @angelobreuer. Are you working on the implementation of WebSockets? Do you need any specific help?

**Comment by angelobreuer** [comment]
Thank you @samirsommaior. I started working today at the implementation of WebSockets on the server-side, which seems to work (https://github.com/angelobreuer/localtunnel.net/commit/58acaac5bea0ac8d15222cbaf2a4bcc73a93fa16). I will add client support for WebSockets now, currently, it only works in passthrough mode (as the client does not forward Upgrade connections correctly). I've tried to make it as abstract as possible, so that not only WebSocket protocols work but all HTTP protocols that use the Upgrade handshake method. I'll let you know if I need any help.
